### ISSUE TITLE
AWS Route53 Provider fails in case of multiple Hosted Zones and wildcard CNAME record

### ISSUE THREAD

**Issue by werbitzky** [janeczku/rancher-letsencrypt#23]
When there are multiple Hosted Zones, rancher-letsencrypt does not use the right enclosing zone in Route53. Instead it seems to pick the first one in the list or maybe there is a bug in the algorithm that tries to find the enclosing Hosted Zone. This results in a failure like below (I've changed the actual domain names).

```
7/19/2016 3:36:27 PMlevel=info msg="Starting Let's Encrypt Certificate Manager v0.3.0 f8965f6"
7/19/2016 3:36:30 PMlevel=info msg="Using locally stored Let's Encrypt account for ****@gmail.com"
7/19/2016 3:36:30 PMlevel=info msg="Trying to obtain SSL certificate (foo.bar.foo.com) from Let's Encrypt CA"
7/19/2016 3:36:31 PMlevel=error msg="[foo.bar.foo.com] Error obtaining certificate: Error presenting token: Failed to change Route 53 record set: InvalidChangeBatch: RRSet with DNS name _acme-challenge.foo.bar.foo.com. is not permitted in zone notfoo.notfoo.com.\n\tstatus code: 400, request id: cd6a6547-4db5-11e6-a875-bf12718e75ac"
```

**Comment by werbitzky** [comment]
Ok, after reading some code, I suspect this is a bug with https://github.com/xenolf/lego
Also it probably happens in the case where we have hosted zones that are a bit similar.

In my case I have two hosted zones like this:

* a.foo.com
* b.foo.com

I try to obtain certs for the domain ```api.b.foo.com``` and the lego library seems to qualify ```a.foo.com``` as the right hosted zone to be used for this even though it is not authoritative for that domain.

**Comment by janeczku** [comment]
@werbitzky Thanks for surfacing these issues. Both of them have to be addressed in the upstream lego library. To address the first issue (arbitrary selection of zones with identical names) the lego `route53` provider needs to take a configuration parameter for the zone-id. The second issue is kind of obscure to me. Might be that there is a bug in the `acme.FindZoneByFqdn()`.
Anyways, can you open a ticket in https://github.com/xenolf/lego for both issues? I'll track the progress here.

**Comment by werbitzky** [comment]
As it turns out, the assumptions were wrong, sorry for that.

```acme.FindZoneByFqdn()``` still seems to be the problem:

I had a _CNAME_ wildcard record in ```b.foo.com``` that was pointing to a domain under ```a.foo.com```. Removing this _CNAME_ record resolved the problem and the right _hosted zone_ was picked.

This still is a bug though and prevents a common usecase. I will open an issue in _lego_ with a complete description and link it here.

**Comment by janeczku** [comment]
@werbitzky I have opened a PR with the fix in the lego repository: https://github.com/xenolf/lego/pull/254

**Comment by alexanderdavidsen** [comment]
I just hit this issue. Any chance of getting this fixed in rancher-letsencrypt? The Lego issue seems to be fixed now.

**Comment by janeczku** [comment]
This should be fixed in v0.4.0. The new template will be available in Rancher catalog soon.

**Comment by kalinon** [comment]
This still seems to be broken in v0.5.0. Found reference to the issue here as well:

https://forums.aws.amazon.com/thread.jspa?messageID=397855&#397855
